Abstract interpretation interim 02.02.2024

LukasPietzschmann opened the pull request.

LukasPietzschmann committed:
Added folder for the best TeX you've ever seen

LukasPietzschmann committed:
Setup poster

LukasPietzschmann committed:
Updated title

LukasPietzschmann committed:
Updated smile

LukasPietzschmann committed:
Updated title

LukasPietzschmann committed:
Added flow-logo repo to provide the logo

LukasPietzschmann committed:
Updated logo submodule

LukasPietzschmann committed:
Structured the poster a bit

LukasPietzschmann committed:
Added section for future work

LukasPietzschmann committed:
I forgot that enumitem is already included lol

LukasPietzschmann committed:
Highlight important points

LukasPietzschmann committed:
Added section about flowR itself

LukasPietzschmann committed:
Rewrote a bit

LukasPietzschmann committed:
Section on flowr and ai

LukasPietzschmann committed:
Rewrote the flowr ai section a bit

LukasPietzschmann committed:
Highlighted important stuff

LukasPietzschmann committed:
Put flowR's flower in a box for easier reuse

LukasPietzschmann committed:
Shorter paragraph titles

LukasPietzschmann committed:
Florian and I = we

LukasPietzschmann committed:
Abstract Interpretation??????

LukasPietzschmann committed:
Updated smile

LukasPietzschmann committed:
Intro example

LukasPietzschmann committed:
Added transparency groups

LukasPietzschmann committed:
Fixed math font

LukasPietzschmann committed:
Put indices also in code

LukasPietzschmann committed:
Better notation of domains

LukasPietzschmann committed:
Reworded step 3

LukasPietzschmann committed:
Fixed math font (again)

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+	\end{tikzpicture}\par
+	\paragraph{Step 2} In the second step happens the actual abstract interpretation.
+	This step will gradually decorate the dataflow graph with the domains of the
+	respective variables.\par
```
"by traversing the CFG"? der kommt bisher auch noch nicht vor.

LukasPietzschmann replied:
Fixed in ad36492

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+	% graph.\par
+	\paragraph{Step 3} The last step uses the decorated dataflow graph to determine and
+	track different properties. We can, e.g., catch endless loops, or, as shown above,
+	eliminate dead paths in the dataflow graph.\par
```
dead paths suggeriert wieder nen CFG - vlt "ignore dead code"?

LukasPietzschmann replied:
Fixed in ad36492

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+		\section*{How abstract interpretation fits into \flowr}
+		\paragraph{\flowr}\flowr{} is a static \emph{dataflow analyzer} and
+		\emph{program slicer} for the R language. Program slicers reduce a given program
+		to a set of statements --- the so called slices --- that influence a variable at
```
technisch sinds auch expressions in R aber das passt :P

LukasPietzschmann replied:
Der Satz beschreibt ja auch Program Slicers allgemein ;)

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+		\paragraph{\flowr}\flowr{} is a static \emph{dataflow analyzer} and
+		\emph{program slicer} for the R language. Program slicers reduce a given program
+		to a set of statements --- the so called slices --- that influence a variable at
+		a specific point in the program --- the slicing criterion. \flowr{} formulates
```
das ist ganz konkret _backward_ slicing :)

LukasPietzschmann replied:
Also _in einer fragenden Tonlage_
> \flowr{} is a static \emph{dataflow analyzer} and \emph{backward program slicer} for the R language.

EagleoutIce replied:
beispielsweise, auch wenn die Implementation von forward jetzt wirklich keine Kunst wäre.

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+		\flowr{} to further \emph{narrow down the slice} of a program. This can be done
+		by removing certain paths from the dataflow graph if we can prove that they will
+		never be executed. Imagine a conditional like \rc{if(i < 10) doit() else run()}.
+		If, through abstract interpretation, we know that \rc{i} is always smaller than
```
den ", through abstract interpretation, " Teil kann man imho auch weglassen, das sollte so +- ne Zeile sparen :D

LukasPietzschmann replied:
Da wir aktuell **noch** kein Platz-Problem haben, würde ich das mal drin lassen (auch weil ich den Teilsatz eig. ganz gern hab xD).
Falls es später eng wird, können wir ihn immer noch entfernen.

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+		never be executed. Imagine a conditional like \rc{if(i < 10) doit() else run()}.
+		If, through abstract interpretation, we know that \rc{i} is always smaller than
+		\rc{10}, there's no need to include the path stemming from the \rc{if}'s
+		else-case in the slice. % TODO: is this okay? Or does it read like we only catch dead code
```
ich finde das passt, wenn du magst kannst du aber ein `e.g.` einweben.

LukasPietzschmann replied:
Ich habs jetzt 20 mal gelesen und bin jetzt gar nicht mehr zufrieden damit xD

Was hältst du von
>  ... This can be done by removing nodes from the dataflow graph that are contained in a provably dead path in the controlflow graph.

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+	meeting. If there's any spare time, Florian often tells me about new
+	\sout{weird}cool things he learned about the R language.
+	\paragraph{Tech Stack} \flowr{} is developed with \emph{TypeScript}, then compiled
+	down to JavaScript with Node.js as its runtime. While we use different libraries for
```
compiled down suggeriert way too much work :P

LukasPietzschmann replied:
Fixed in ad36492

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+	\paragraph{Tech Stack} \flowr{} is developed with \emph{TypeScript}, then compiled
+	down to JavaScript with Node.js as its runtime. While we use different libraries for
+	utilitarian tasks --- like chai for assertions, mocha for tests, or tslog for
+	logging --- all major functionalities are implemented \emph{by hand}. This includes
```
vlt noch ein mini kommentar darüber warum von hand :D weil es für die R-Sprache (quasi) nix (auf jeden Fall nichts vergleichbares) gibt.

LukasPietzschmann replied:
Fixed in ad36492

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+	\paragraph{QA} To ensure the employment of best practices, we always do \emph{code
+	reviews} on pull requests. This massively helps with keeping the code easy to
+	understand and well readable.\par
+	If an error slips through during the review, rest assured that one of \flowr's
```
"rest assured" transportiert mehr Vertrauen als ich in die ganze code base habe :P

LukasPietzschmann replied:
Fixed in ad36492

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+	extensive suite of over \emph{1000 tests} will certainly catch it. Additionally, we
+	consistently introduce new tests promptly whenever there is new code in the
+	project.\par
+	We also make heavy use of \emph{assertions} --- or how we call them:~guards ---
```
man könnte auch sagen contracts - das war zmd. was ich ursprünglich machen wollte :D

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+\uni{Ulm University}
+\institute{Institute of Software Engineering and Programming Languages}
+\date{02/02/2024}
+\logo{logos/splogo.png}
```
soll das so bleiben? wenn ja, dann könnte man vlt noch dafür sorgen, dass sich die Höhe/Position an irgendetwas orientiert oder es vlt auch aussieht wie ne bookmark?

LukasPietzschmann replied:
Ich bin mir nicht ganz sicher, was du mit der ersten Hälfte des zweiten Satzes meinst. Der Kasten, in dem das Logo ist, ist ja nicht _random_ platziert. Und das Logo im Kasten ist einfach mittig.

Aber wir können gerne ausprobieren, wie ein Lesezeichen aussieht :)

LukasPietzschmann replied:
75101fa hat das Lesezeichen hinzugefügt

EagleoutIce commented on 2324-wt/abstract-interpretation/interim/poster.tex:
```diff
@@ -0,0 +1,221 @@
+	\newsavebox\exampleb
+	\begin{lrbox}\exampleb
+	\begin{lstlisting}[language=R,style=firastylep]
+x§\j1§ <- 10
```
wie direkt angemerkt :D (`§§` vs. `§`)

LukasPietzschmann replied:
Fixed in 53a5596

EagleoutIce commented:
Man könnte vlt erwähnen, was der bisherige Implementationsstand ist (das ist dann vermutlich Teil von Future Work) bzw. auf constant folding & constant propagation linken.

LukasPietzschmann committed:
Updated smile

LukasPietzschmann committed:
Added fira fonts to the repo

LukasPietzschmann committed:
Reworded stuff

LukasPietzschmann committed:
Made SP logo to a bookmark

LukasPietzschmann committed:
Changed opacity a bit

LukasPietzschmann committed:
Fixed overfull boxes

LukasPietzschmann committed:
Added example

LukasPietzschmann committed:
Interval -> domain

LukasPietzschmann committed:
Light blue tint for boxes

LukasPietzschmann committed:
Changed wording

LukasPietzschmann committed:
Switched domain representation

LukasPietzschmann committed:
Moved arrows before nodes

LukasPietzschmann committed:
Consistently smaller line width

LukasPietzschmann committed:
let is now sf not tt

LukasPietzschmann committed:
Only one makeatletter

LukasPietzschmann committed:
Uniform height of example boxes

LukasPietzschmann committed:
Make line numbers more readable

LukasPietzschmann committed:
Use \left and \right for intervals

LukasPietzschmann committed:
Added line 0 in table

LukasPietzschmann committed:
Added 7th line to the example

LukasPietzschmann committed:
Restore parskip in minipage

LukasPietzschmann committed:
First text draft

LukasPietzschmann committed:
Future work

LukasPietzschmann committed:
Highlighted important stuff

LukasPietzschmann committed:
Reworded the text a bit

LukasPietzschmann committed:
Spacing

LukasPietzschmann committed:
First round of rewrites

LukasPietzschmann committed:
Second round of rewrites

LukasPietzschmann committed:
Rewrites #4 (i feel like i'm playing a game of whack-a-mole)

This commit message was generated by GitHub copilot, who seems to like
games of whack-a-mole. I'm not sure if i should be worried or not.
GH copilot also generated this message and said that it is not sure if
it should be worried or not. I'm not sure if i should be worried or not.

LukasPietzschmann committed:
Small changes

LukasPietzschmann committed:
Bigger footer

LukasPietzschmann committed:
Rewordings

LukasPietzschmann committed:
RIP SP

LukasPietzschmann committed:
Spacing

LukasPietzschmann committed:
flowr power
